<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.11"/>
<title>libsequence: Alphabets defined in libsequence</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
  $(window).load(resizeHeight);
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">libsequence
   &#160;<span id="projectnumber">1.9.1</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.11 -->
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li class="current"><a href="modules.html"><span>Modules</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li><a href="examples.html"><span>Examples</span></a></li>
    </ul>
  </div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
$(document).ready(function(){initNavTree('group__Alphabets.html','');});
</script>
<div id="doc-content">
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#typedef-members">Typedefs</a> &#124;
<a href="#func-members">Functions</a> &#124;
<a href="#var-members">Variables</a>  </div>
  <div class="headertitle">
<div class="title">Alphabets defined in libsequence</div>  </div>
</div><!--header-->
<div class="contents">

<p>Various character sets for different types of analysis.  
<a href="#details">More...</a></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structSequence_1_1ambiguousNucleotide.html">Sequence::ambiguousNucleotide</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Tests if a character is in the set A,G,C,T.  <a href="structSequence_1_1ambiguousNucleotide.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structSequence_1_1invalidPolyChar.html">Sequence::invalidPolyChar</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">This functor can be used to determine if a range contains characters that the SNP analysis routines in this library cannot handle gracefully.  <a href="structSequence_1_1invalidPolyChar.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="typedef-members"></a>
Typedefs</h2></td></tr>
<tr class="memitem:ga30379e60c1fe1fd37d1703e346151230"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ga30379e60c1fe1fd37d1703e346151230"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__Alphabets.html#ga30379e60c1fe1fd37d1703e346151230">Sequence::alphabet_t</a> = std::array&lt; const char, 16 &gt;</td></tr>
<tr class="memdesc:ga30379e60c1fe1fd37d1703e346151230"><td class="mdescLeft">&#160;</td><td class="mdescRight">Container type for nucleotide alphabets. <br /></td></tr>
<tr class="separator:ga30379e60c1fe1fd37d1703e346151230"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:ga04b131da5a173bc96dff2079b18e5ae6"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__Alphabets.html#ga04b131da5a173bc96dff2079b18e5ae6">Sequence::isDNA</a> (const char &amp;ch)</td></tr>
<tr class="memdesc:ga04b131da5a173bc96dff2079b18e5ae6"><td class="mdescLeft">&#160;</td><td class="mdescRight">test if character is part of <a class="el" href="group__Alphabets.html#ga790250a2ef0192d63a54b94ef646be32" title="Alphabet for DNA sequences Valid DNA characters. Upper-case only. Only - is accepted as gap character...">Sequence::dna_alphabet</a>  <a href="group__Alphabets.html#ga04b131da5a173bc96dff2079b18e5ae6">More...</a><br /></td></tr>
<tr class="separator:ga04b131da5a173bc96dff2079b18e5ae6"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="var-members"></a>
Variables</h2></td></tr>
<tr class="memitem:ga790250a2ef0192d63a54b94ef646be32"><td class="memItemLeft" align="right" valign="top">const alphabet_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__Alphabets.html#ga790250a2ef0192d63a54b94ef646be32">Sequence::dna_alphabet</a></td></tr>
<tr class="memdesc:ga790250a2ef0192d63a54b94ef646be32"><td class="mdescLeft">&#160;</td><td class="mdescRight">Alphabet for DNA sequences Valid DNA characters. Upper-case only. Only - is accepted as gap characters.  <a href="group__Alphabets.html#ga790250a2ef0192d63a54b94ef646be32">More...</a><br /></td></tr>
<tr class="separator:ga790250a2ef0192d63a54b94ef646be32"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga72312cc3c51db6950ecdf31ffaf09159"><td class="memItemLeft" align="right" valign="top">const alphabet_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__Alphabets.html#ga72312cc3c51db6950ecdf31ffaf09159">Sequence::dna_poly_alphabet</a></td></tr>
<tr class="memdesc:ga72312cc3c51db6950ecdf31ffaf09159"><td class="mdescLeft">&#160;</td><td class="mdescRight">Alphabet for polymorphism (SNP) analysis. 16 characters are used so that we may encode 2 nucleotides in a 8-bit integer.  <a href="group__Alphabets.html#ga72312cc3c51db6950ecdf31ffaf09159">More...</a><br /></td></tr>
<tr class="separator:ga72312cc3c51db6950ecdf31ffaf09159"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa15be372eac7115b3fe355c57cb9041b"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="gaa15be372eac7115b3fe355c57cb9041b"></a>
const alphabet_t::size_type&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__Alphabets.html#gaa15be372eac7115b3fe355c57cb9041b">Sequence::NOTPOLYCHAR</a> = dna_poly_alphabet.size()</td></tr>
<tr class="memdesc:gaa15be372eac7115b3fe355c57cb9041b"><td class="mdescLeft">&#160;</td><td class="mdescRight">An index from dna_poly_alphabet &gt;= this is not a valid character for variation analysis. <br /></td></tr>
<tr class="separator:gaa15be372eac7115b3fe355c57cb9041b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8dc73f7a5d4039193d66e4020d65ca95"><td class="memItemLeft" align="right" valign="top">const alphabet_t::size_type&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__Alphabets.html#ga8dc73f7a5d4039193d66e4020d65ca95">Sequence::POLYEOS</a></td></tr>
<tr class="memdesc:ga8dc73f7a5d4039193d66e4020d65ca95"><td class="mdescLeft">&#160;</td><td class="mdescRight">The value of terminating an encoded string of SNP data.  <a href="group__Alphabets.html#ga8dc73f7a5d4039193d66e4020d65ca95">More...</a><br /></td></tr>
<tr class="separator:ga8dc73f7a5d4039193d66e4020d65ca95"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<p>Various character sets for different types of analysis. </p>
<h2 class="groupheader">Function Documentation</h2>
<a class="anchor" id="ga04b131da5a173bc96dff2079b18e5ae6"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool Sequence::isDNA </td>
          <td>(</td>
          <td class="paramtype">const char &amp;&#160;</td>
          <td class="paramname"><em>ch</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>test if character is part of <a class="el" href="group__Alphabets.html#ga790250a2ef0192d63a54b94ef646be32" title="Alphabet for DNA sequences Valid DNA characters. Upper-case only. Only - is accepted as gap character...">Sequence::dna_alphabet</a> </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">ch</td><td>Character to test </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>true if ch is in <a class="el" href="group__Alphabets.html#ga790250a2ef0192d63a54b94ef646be32" title="Alphabet for DNA sequences Valid DNA characters. Upper-case only. Only - is accepted as gap character...">Sequence::dna_alphabet</a>, false otherwise </dd></dl>
<dl class="section note"><dt>Note</dt><dd>case-insensitive via std::toupper </dd></dl>

<p>Definition at line <a class="el" href="SeqAlphabets_8cc_source.html#l00025">25</a> of file <a class="el" href="SeqAlphabets_8cc_source.html">SeqAlphabets.cc</a>.</p>

</div>
</div>
<h2 class="groupheader">Variable Documentation</h2>
<a class="anchor" id="ga790250a2ef0192d63a54b94ef646be32"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">const alphabet_t Sequence::dna_alphabet</td>
        </tr>
      </table>
</div><div class="memdoc">
<b>Initial value:</b><div class="fragment"><div class="line">{ {<span class="charliteral">&#39;A&#39;</span>,<span class="charliteral">&#39;C&#39;</span>,<span class="charliteral">&#39;G&#39;</span>,<span class="charliteral">&#39;T&#39;</span>,</div><div class="line">        <span class="charliteral">&#39;R&#39;</span>,<span class="charliteral">&#39;Y&#39;</span>,<span class="charliteral">&#39;S&#39;</span>,<span class="charliteral">&#39;W&#39;</span>,</div><div class="line">        <span class="charliteral">&#39;K&#39;</span>,<span class="charliteral">&#39;M&#39;</span>,<span class="charliteral">&#39;B&#39;</span>,<span class="charliteral">&#39;D&#39;</span>,</div><div class="line">        <span class="charliteral">&#39;H&#39;</span>,<span class="charliteral">&#39;V&#39;</span>,<span class="charliteral">&#39;N&#39;</span>,<span class="charliteral">&#39;-&#39;</span>} }</div></div><!-- fragment -->
<p>Alphabet for DNA sequences Valid DNA characters. Upper-case only. Only - is accepted as gap characters. </p>
<dl class="section note"><dt>Note</dt><dd><a href="http://www.bioinformatics.org/sms/iupac.html,">http://www.bioinformatics.org/sms/iupac.html,</a> excluding U, and ., which is redundant with - </dd></dl>

<p>Definition at line <a class="el" href="Sequence_2SeqAlphabets_8hpp_source.html#l00029">29</a> of file <a class="el" href="Sequence_2SeqAlphabets_8hpp_source.html">Sequence/SeqAlphabets.hpp</a>.</p>

</div>
</div>
<a class="anchor" id="ga72312cc3c51db6950ecdf31ffaf09159"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">const alphabet_t Sequence::dna_poly_alphabet</td>
        </tr>
      </table>
</div><div class="memdoc">
<b>Initial value:</b><div class="fragment"><div class="line">{ {<span class="charliteral">&#39;A&#39;</span>,<span class="charliteral">&#39;C&#39;</span>,<span class="charliteral">&#39;G&#39;</span>,<span class="charliteral">&#39;T&#39;</span>, </div><div class="line">        <span class="charliteral">&#39;0&#39;</span>,<span class="charliteral">&#39;1&#39;</span>,<span class="charliteral">&#39;-&#39;</span>,<span class="charliteral">&#39;N&#39;</span>, </div><div class="line">        <span class="charliteral">&#39;\0&#39;</span>, </div><div class="line">        } }</div></div><!-- fragment -->
<p>Alphabet for polymorphism (SNP) analysis. 16 characters are used so that we may encode 2 nucleotides in a 8-bit integer. </p>

<p>Definition at line <a class="el" href="Sequence_2SeqAlphabets_8hpp_source.html#l00037">37</a> of file <a class="el" href="Sequence_2SeqAlphabets_8hpp_source.html">Sequence/SeqAlphabets.hpp</a>.</p>

</div>
</div>
<a class="anchor" id="ga8dc73f7a5d4039193d66e4020d65ca95"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">const alphabet_t::size_type Sequence::POLYEOS</td>
        </tr>
      </table>
</div><div class="memdoc">
<b>Initial value:</b><div class="fragment"><div class="line">= alphabet_t::size_type( std::distance(<a class="code" href="group__Alphabets.html#ga72312cc3c51db6950ecdf31ffaf09159">dna_poly_alphabet</a>.begin(),</div><div class="line">                                                                             std::find(<a class="code" href="group__Alphabets.html#ga72312cc3c51db6950ecdf31ffaf09159">dna_poly_alphabet</a>.begin(),</div><div class="line">                                                                                       <a class="code" href="group__Alphabets.html#ga72312cc3c51db6950ecdf31ffaf09159">dna_poly_alphabet</a>.end(),</div><div class="line">                                                                                       <span class="charliteral">&#39;\0&#39;</span>)</div><div class="line">                                                                             ) )</div><div class="ttc" id="group__Alphabets_html_ga72312cc3c51db6950ecdf31ffaf09159"><div class="ttname"><a href="group__Alphabets.html#ga72312cc3c51db6950ecdf31ffaf09159">Sequence::dna_poly_alphabet</a></div><div class="ttdeci">const alphabet_t dna_poly_alphabet</div><div class="ttdoc">Alphabet for polymorphism (SNP) analysis. 16 characters are used so that we may encode 2 nucleotides ...</div><div class="ttdef"><b>Definition:</b> <a href="Sequence_2SeqAlphabets_8hpp_source.html#l00037">Sequence/SeqAlphabets.hpp:37</a></div></div>
</div><!-- fragment -->
<p>The value of terminating an encoded string of SNP data. </p>

<p>Definition at line <a class="el" href="Sequence_2SeqAlphabets_8hpp_source.html#l00050">50</a> of file <a class="el" href="Sequence_2SeqAlphabets_8hpp_source.html">Sequence/SeqAlphabets.hpp</a>.</p>

</div>
</div>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="footer">Generated on Wed Apr 19 2017 09:49:52 for libsequence by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.11 </li>
  </ul>
</div>
</body>
</html>
